### PR TITLE
Control destroy

### PR DESCRIPTION
**Pull request by dorileo**
It was initially part of PR #87

### CODE REVIEW

**Comment by VladimirMarkelov** [comment]
I see your point. I think if you have chosen Frame to add Destroy method as an example it would have been better :)  Anyway, are you going to add Destroy for other controls? For Label Destroy method is hardly useful since it rarely has children.

**Comment by dorileo** [comment]
Hi @VladimirMarkelov 

I don't think Destroy() should care about the destroyed children but only taking it out of its parents children chain. I'm still wondering if we should remove the destroyed control's children - for now my answer to that is no.

I would like to have all controls implementing Destroy(), I understand it's very useful to any of them. In that sense I did a "common" implementation between all controls in the BaseControl level, but given the go's comparison operations limitation, where comparisons between interface and non-interface values (namely between BaseControl and Control) will never match the being destroyed control against the object's children slice, to fix that I came up with a implementation of a NewBaseControl() to be issued wen allocating a new control (i.e new(Window)), that MakeBaseControl() will set a reference id to each control, so we can compare BaseControl against Control.

In the control level it would simply become:

```go
wnd := new(Window)
wnd.BaseControl = NewBaseControl()
```

The implementation for NewBaseControl() is as simple as:

```go
func NewBaseControl() BaseControl {
	idMtx.Lock()
	defer idMtx.Unlock()

	idRefs = idRefs + 1
	return BaseControl{refID: idRefs}
}
```

and the RemoveChild() looks like this (I also changed it to non exported and the exported interface is only Destroy()):

```go
func (c *BaseControl) removeChild(control Control) {
	children := []Control{}

	for _, child := range c.children {
		if child.RefID() == control.RefID() {
			continue
		}

		children = append(children, child)
	}
	c.children = nil

	for _, child := range children {
		c.AddChild(child)
	}
}
```

note that the Control interface had the addition of a RefID() method to compare ids.

And for completeness sake, here's Destroy() implementation:

```go
func (c *BaseControl) Destroy() {
	c.parent.removeChild(c)
}
```


Anyway, long story short, we can implement it in the BaseControl level but will have the refID inconvenience - I can show you the resulting patches or update this pull request if you think this could be a good way to go.

Another option would be to simply implement Destroy() on each control, but we would end up having similar Destroy() methods everywhere :(

**Comment by VladimirMarkelov** [comment]
I thought about having something like ID for controls but since it worked without it I have not implemented anything. Instead of heavy mutex we can go with atomic int64, and provide a global private function like:
```
func nextRefId() int64 {
   return atomic.AddInt64(&globalRefId, 1)  
}
```
As for inconvenience, I agree: we have to set this RefId for all controls (but it is only internal one, so an end-user won't see it and won't feel the difference), and public interface method `RefId()` that is not a big trouble. People can use it for their internal stuff(but I am not sure for what ones  yet :) ).

**Comment by dorileo** [comment]
Hi @VladimirMarkelov I pushed a new version with the refID version we talked about last week.

Thanks...
